pkg/storegateway/sharding_strategy_test.go

```diff
@@ -499,3 +500,345 @@ type shardingLimitsMock struct {
+	generateTokens := func(instanceIndex, tokensPerInstance int) []uint32 {
+		tokens := make([]uint32, tokensPerInstance)
+		baseToken := uint32(instanceIndex * 1000000)
+		for i := 0; i < tokensPerInstance; i++ {
+			tokens[i] = baseToken + uint32(i*10000)
+		}
+		return tokens
+	}
+
+	tokensPerInstance := 128
+
+	// Helper to create instance definitions for a zone.
+	type instanceDef struct {
+		id     string
+		addr   string
+		zone   string
+		tokens []uint32
+	}
+
+	createZoneInstances := func(zone string, startIndex int) []instanceDef {
+		instances := make([]instanceDef, instancesPerZone)
+		for i := 0; i < instancesPerZone; i++ {
+			instances[i] = instanceDef{
+				id:     fmt.Sprintf("store-gateway-%s-%d", zone, i+1),
+				addr:   fmt.Sprintf("%s-%d.example.com", zone, i+1),
+				zone:   zone,
+				tokens: generateTokens(startIndex+i, tokensPerInstance),
+			}
+		}
+		return instances
+	}
+
+	// Create initial instances for 3 zones.
+	zoneAInstances := createZoneInstances("zone-a", 0)
+	zoneBInstances := createZoneInstances("zone-b", instancesPerZone)
+	zoneCInstances := createZoneInstances("zone-c", instancesPerZone*2)
+
+	// Helper type to capture state of an instance.
+	// We store blocks per-user to ensure we're tracking the exact same block ownership,
+	// not just aggregate counts which could mask redistribution issues.
+	type instanceState struct {
+		users         []string
+		blocksPerUser map[string][]ulid.ULID // userID -> blocks owned for that user
+	}
+
+	// Helper to capture state for all instances.
+	captureState := func(t *testing.T, r *ring.Ring, instances []instanceDef, replicationFactor int, limits ShardingLimits) map[string]instanceState {
+		t.Helper()
+		state := make(map[string]instanceState)
+
+		for _, inst := range instances {
+			strategy := NewShuffleShardingStrategy(r, inst.id, inst.addr, NewNopDynamicReplication(replicationFactor), limits, log.NewNopLogger())
+
+			// Capture FilterUsers result.
+			filteredUsers, err := strategy.FilterUsers(ctx, userIDs)
+			require.NoError(t, err, "FilterUsers failed for instance %s", inst.id)
+
+			// Capture FilterBlocks result for each user separately.
+			blocksPerUser := make(map[string][]ulid.ULID)
+			for _, userID := range userIDs {
+				metas := make(map[ulid.ULID]*block.Meta, len(blocks))
+				for _, blockID := range blocks {
+					metas[blockID] = &block.Meta{
+						BlockMeta: tsdb.BlockMeta{
+							ULID:    blockID,
+							MinTime: time.Now().Add(-24 * time.Hour).UnixMilli(),
+							MaxTime: time.Now().UnixMilli(),
+						},
+					}
+				}
+
+				synced := extprom.NewTxGaugeVec(nil, prometheus.GaugeOpts{}, []string{"state"})
+				err := strategy.FilterBlocks(ctx, userID, metas, nil, synced)
+				require.NoError(t, err, "FilterBlocks failed for instance %s, user %s", inst.id, userID)
+
+				var userBlocks []ulid.ULID
+				for blockID := range metas {
+					userBlocks = append(userBlocks, blockID)
```

Comment:
[nit] You can simply append to `blocksPerUser`. There's no need of having `userBlocks`.